projects/firmware_homework/src/main.c

```diff
@@ -0,0 +1,46 @@
+// Standard library includes
+#include <stdint.h>
+#include <stdio.h>
+// Midnight Sun includes
+#include "delay.h"
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+#include "wait.h"
+// Macros
+#define DELAY_TIME 500
+
+// Structure
+typedef struct Counters {
+  uint8_t counter_a;
+  uint8_t counter_b;
+} Counters;
+
+// Private function
+static void prv_timer_callback(SoftTimerId timer_id, void *context) {
+  Counters *storage = context;
+  // Counter A needs to be updated twice
+  storage->counter_a++;
+  LOG_DEBUG("Counter A: %i\n", storage->counter_a);
+  storage->counter_a++;
+  LOG_DEBUG("Counter A: %i\n", storage->counter_a);
+  // Counter B needs to be updated once
+  storage->counter_b++;
+  LOG_DEBUG("Counter B: %i\n", storage->counter_b);
+  // Start the timer again
+  soft_timer_start_millis(DELAY_TIME, prv_timer_callback, storage, NULL);
```

Comment:
The homework does say:
> increment `counter_a` every half a second, and `counter_b` every second

This has the right output, but it increments `counter_a` *twice* every half second and `counter_b` once every half second.